utils/s3/credentials_providers/sts_assume_role_credentials_provider.cc

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "sts_assume_role_credentials_provider.hh"
+
+#include "utils/UUID.hh"
+#include "utils/http.hh"
+#include "utils/s3/client.hh"
+#include <rapidxml.h>
+#include <seastar/core/coroutine.hh>
+#include <seastar/http/request.hh>
+#include <seastar/util/short_streams.hh>
+
+namespace aws {
+
+static logging::logger sts_logger("sts");
+
+sts_assume_role_credentials_provider::sts_assume_role_credentials_provider(const std::string& _host, unsigned _port, bool _is_secured)
+    : sts_host(_host), port(_port), is_secured(_is_secured) {
+}
+
+sts_assume_role_credentials_provider::sts_assume_role_credentials_provider(const std::string& _region, const std::string& _role_arn)
+    : sts_host(seastar::format("sts.{}.amazonaws.com", _region)), role_arn(_role_arn) {
+}
+
+future<s3::aws_credentials> sts_assume_role_credentials_provider::get_aws_credentials() {
+    co_await reload();
+    co_return creds;
+}
+
+bool sts_assume_role_credentials_provider::is_time_to_refresh() const {
+    return std::chrono::system_clock::now() >= creds.expires_at;
+}
+
+future<> sts_assume_role_credentials_provider::reload() {
+    if (!is_time_to_refresh() || !creds) {
+        co_return co_await update_credentials();
+    }
+}
+
+future<> sts_assume_role_credentials_provider::update_credentials() {
+    auto req = http::request::make("POST", sts_host, "/");
+    // Just set this version
+    // https://github.com/aws/aws-sdk-cpp/blob/8d68be52dcad85095753e069a4355e241f1edb1c/generated/src/aws-cpp-sdk-sts/source/model/AssumeRoleRequest.cpp#L143
+    req.query_parameters["Version"] = "2011-06-15";
+    req.query_parameters["DurationSeconds"] = format("{}", session_duration);
+    req.query_parameters["Action"] = "AssumeRole";
+    req.query_parameters["RoleSessionName"] = format("{}", utils::make_random_uuid());
+    req.query_parameters["RoleArn"] = role_arn;
```

Comment:
Will it work if role_arn is empty string?

Reply:
It will "work" as expected, failing to obtain creds